Support Groups as Subject Kind for Tenant Namespace RoleBindings created by Capsule

MaxFedotov opened the pull request:
This PR closes #70  and adds additional configuration parameter `use-groups-as-tenant-owner` which allows Tenant Namespace RoleBindings to be created with `"Kind": "Group"`

It's completely backward-compatible, as default is set to `false`.

bsctl commented:
@MaxFedotov thanks for your interest in Capsule and for pushing this PR.

Should we consider to have a more general approach for the Tenant Ownership.
What about to leave the cluster admin to set the `Subject.Kind` of the owner by tenant instead of a general setting?

I mean something like this:

```yaml
apiVersion: capsule.clastix.io/v1alpha1
kind: Tenant
metadata:
  name: oil
spec:
...
 subjects owner:
   kind: Group
   name: foo
```

Where the `kind` can be `User`, `Group`, and `ServiceAccount` too. The latter case can be useful if the tenant owner should be an application.

Not sure if technically feasible. @prometherion @MaxFedotov what do you think?

prometherion commented:
The main concern I have is upon Namespace creation, since we're using an indexer to speed up the Owner lookup at webhook level.

I need to investigate and check if it's feasible or not, since we have to lookup also per `Kind` rather than just the name.

MaxFedotov commented:
@prometherion 
You are talking about this code - https://github.com/clastix/capsule/blob/0f935d53b7acb17379e749a2d64e2f631adb6779/pkg/webhook/owner_reference/patching.go#L84?

prometherion commented:
Exactly, that's a field indexer provided by the related struct

https://github.com/clastix/capsule/blob/0f935d53b7acb17379e749a2d64e2f631adb6779/pkg/indexer/tenant/owner.go#L37:L42

MaxFedotov commented:
@prometherion 
> I need to investigate and check if it's feasible or not, since we have to lookup also per `Kind` rather than just the name.

There is a way to do it - https://github.com/kubernetes-sigs/controller-runtime/issues/612

in our case we can use function like
```
func getValue(tenant) string {
   return tenant.Spec.SubjectOwner.Kind+":"+tenant.Spec.SubjectOwner.Name
}
```

But there is another problem, when https://github.com/clastix/capsule/blob/0f935d53b7acb17379e749a2d64e2f631adb6779/pkg/webhook/owner_reference/patching.go#L84-L93
is called, there is no way to get from `req admission.Request`an information about a `Kind` of TenantOwner. 
So first we need get TenantList for `User:req.UserInfo.Groups` and then if `len(req.UserInfo.Groups) > 0` loop through it and for each group try to find TenantList which match `Group:group_name` and append them.

Another way, is to enable by default `force-tenant-prefix`, which will make looking for TenantList match easier, as we can just get tenant name from Namespace name :)

MaxFedotov commented:
@prometherion 
So what i've done in last commit:
- Updated Capsule CRD to add support for new struct 
```
type OwnerSpec struct {
	Name string `json:"name"`
	Kind Kind   `json:"kind"`
}
```
- Rewritten indexer to support complex Owner field like 
```
tenant.Spec.Owner.Kind.String() + ":" + tenant.Spec.Owner.Name
```
- Updated owner_reference webhook with a logic, that I proposed in previous comment
> if force-tenant-prefix is enabled, extract tenantName from Namespace name and do simple lookup.
> if not - find first matching tenant either from User or from Group kind
- Added tenant name validation webhook in order to restrict tenant names to only alphanumeric characters, so we will be able to extract tenant from namespace name in owner_reference webhook
- Created new tests and updated existing to support these changes
- Updated docs

prometherion commented:
Thanks @MaxFedotov, can't wait for take a look to it.

@bsctl do you think we should version a new API since this breaking change in the contract?

IMO, no need since we haven't yet released a stable version :)

prometherion commented on e2e/tenant_name_webhook_test.go:
```diff
@@ -0,0 +1,53 @@
+	"github.com/clastix/capsule/api/v1alpha1"
+)
+
+var _ = Describe("creating a Tenant with wrong name", func() {
```
I guess the same here: this test shouldn't be related to this PR, right?

MaxFedotov replied:
This is a test for a webhook, I've had to add in order to be able to extract tenant name from namespace name

prometherion commented on e2e/tenant_resources_test.go:
```diff
@@ -27,7 +27,7 @@ import (
 	. "github.com/onsi/gomega"
 	corev1 "k8s.io/api/core/v1"
-	"k8s.io/api/networking/v1"
+	v1 "k8s.io/api/networking/v1"
```
Please, use the alias `networkingv1`

prometherion replied:
these imports are duplicated, please jus use `networkingv1` and drop `v1 "k8s.io/api/networking/v1"`

MaxFedotov replied:
sorry, seems like my VSCode is adding this. Fixed

prometherion commented on e2e/utils_test.go:
```diff
@@ -23,15 +23,18 @@ import (
 
 const (
-	defaultTimeoutInterval = 10 * time.Second
+	defaultTimeoutInterval = 30 * time.Second
```
Why incrementing to 30 secs?

If this is required for the Pod restart, maybe it's required to hack just the expectation.

MaxFedotov replied:
fixed

prometherion commented on pkg/webhook/owner_reference/patching.go:
```diff
@@ -81,17 +85,42 @@ func (r *handler) OnCreate(ctx context.Context, req admission.Request, clt clien
-	}); err != nil {
-		return admission.Errored(http.StatusBadRequest, err)
+	// assigning namespace to Tenant in case of --force-tenant-prefix flag enabled
+	if r.forceTenantPrefix {
```
Do we really need this? We already have the `tenant_prefix` webhook.

MaxFedotov replied:
As i understand, `tenant_prefix` is just validating webhook. 
I've modified logic in this webhook to work like this:
- if force-tenant-prefix is enabled, extract tenantName from Namespace name and do simple lookup.
- if not - find first matching tenant either from User or from Group kind

bsctl replied:
As I understand, if `--force-tenant-prefix` is true, the name of the tenant is already in the namespace, i.e. `tenant-namespace` so do a simple lookup in that case.

If `--force-tenant-prefix` is false, we have to do a more complex lookup to find which tenant the namespace belong to.

MaxFedotov commented on e2e/ingress_class_test.go:
```diff
@@ -38,10 +38,13 @@ import (
 	tnt := &v1alpha1.Tenant{
 		ObjectMeta: metav1.ObjectMeta{
-			Name: "ingress-class",
+			Name: "ingressclass",
```
I've changed a logic of owner_reference following way:
- if force-tenant-prefix is enabled, extract tenantName from Namespace name and do simple lookup.
- if not - find first matching tenant either from User or from Group kind

In order to be ably to extract tenantName from Namespace when `force-tenant-prefix` is enabled, I've had to create additional webhook, which limits TenantName to only alphanumeric characters, because without it, it would be impossible to get tenant name from namespace name (e.g. if we allow dashes in tenant name, how we would extract tenant name from a namespace named like 'bla-bla-1'?)

So I had to change tenant names in all tests.

MaxFedotov commented on e2e/force_tenant_prefix_test.go:
```diff
@@ -0,0 +1,91 @@
+	"github.com/clastix/capsule/api/v1alpha1"
+)
+
+var _ = Describe("creating a Namespace with --force-tenant-name flag", func() {
```
As i'm relying on `force-tenant-name` hook in order to set owner reference, I had to verify that it's working, so I've added additional test for it :)

MaxFedotov commented on e2e/selecting_tenant_no_label_test.go:
```diff
@@ -0,0 +1,99 @@
+	"github.com/clastix/capsule/api/v1alpha1"
+)
+
+var _ = Describe("creating a Namespace without a Tenant selector", func() {
```
The same, as previous. As i've modified `owner_reference` webhook logic, I've had to verify that this logic is working

prometherion reviewed (changes requested):
Some typos and more details requests

MaxFedotov commented on pkg/webhook/tenant_name/validating.go:
```diff
@@ -0,0 +1,71 @@
+
+	matched, _ := regexp.MatchString(`^[a-z0-9]([a-z0-9]*[a-z0-9])?$`, tnt.GetName())
+	if !matched {
+		fmt.Printf("%s not mached", tnt.GetName())
```
fixed

MaxFedotov commented on pkg/webhook/tenant_name/validating.go:
```diff
@@ -0,0 +1,71 @@
+	matched, _ := regexp.MatchString(`^[a-z0-9]([a-z0-9]*[a-z0-9])?$`, tnt.GetName())
+	if !matched {
+		fmt.Printf("%s not mached", tnt.GetName())
+		return admission.Denied("Tenant name had forbidden characters")
```
fixed

MaxFedotov commented on pkg/webhook/tenant_name/validating.go:
```diff
@@ -0,0 +1,71 @@
+	}
+	fmt.Printf("%s", tnt.GetName())
+
+	matched, _ := regexp.MatchString(`^[a-z0-9]([a-z0-9]*[a-z0-9])?$`, tnt.GetName())
```
So we are adding an ability to specify `Kind` of tenant owner. 
Current implementation of owner_refrence webhook use following logic to set tenant for a namespace
https://github.com/clastix/capsule/blob/a99153cbe70432d7e21ad5a2610ba737b1ba27d3/pkg/webhook/owner_reference/patching.go#L84-L93
So basically we find all Tenants for a user\group who created a namespace, and select first one from it (and I think that this logic should be rewritten if future because it's a bit strange for me - a good option will be to set `force-tenant-prefix` by default, so all this magic will gone)

But then we had such thing as `force-tenant-prefix` flag. When this flag is set, we can simply deduct tenant name from namespace name, and that is done in a following code in owner_reference webhook:
https://github.com/clastix/capsule/blob/9d845b367852fb19d1e2756fe7e27fd6338dcdf1/pkg/webhook/owner_reference/patching.go#L89-L102

But here another problem arises. If `-` is used in a tenant name, we won't be able to do it, we need to had some complicated logic with multiple splits and lookups, until we find a suitable tenant. So it order to simplify this process, I've added a webhook which enforces Tenant name to use only alphanumeric characters, thus simplifying overall logic

MaxFedotov replied:
#76 created

MaxFedotov commented on pkg/webhook/router.go:
```diff
@@ -54,8 +54,9 @@ type handlerRouter struct {
-	if !utils.UserGroupList(req.UserInfo.Groups).IsInCapsuleGroup() {
-		// not a Capsule user, can be skipped
+
+	if !utils.UserGroupList(req.UserInfo.Groups).IsInCapsuleGroup() && req.Kind.Kind != "Tenant" {
```
Because we need `tenant_name` webhook to always fire. Tenants can be created by admins, who can be not a members of `capsule-user-group`. Without this condition they will be able to create tenants, whose name is not validated by `tenant_name` webhook

prometherion replied:
Oh, I see... it seems we need a sort of Middleware here.
Sorry to block your PR here but we need to figure out a better webhook middleware piping.

MaxFedotov replied:
No problem :) For example, we can pass here
https://github.com/clastix/capsule/blob/9d845b367852fb19d1e2756fe7e27fd6338dcdf1/pkg/webhook/router.go#L51-L56
and additional map of `enforcedResourceKinds` with the Kinds of resources we want hooks to always fire for
e.g
```
enforcedResourceKinds := map[string]bool {
    "Tenant": true,
}
```
and here https://github.com/clastix/capsule/blob/9d845b367852fb19d1e2756fe7e27fd6338dcdf1/main.go#L141

use 
```
err = webhook.Register(mgr, capsuleGroup,enforcedResourceKinds, wl...)
```
so the code will be like 
```

if !utils.UserGroupList(req.UserInfo.Groups).IsInCapsuleGroup(r.capsuleGroup) && !enforcedResourceKinds[req.Kind.Kind]  {
		// not a Capsule user, not a tenant kind, can be skipped
		return admission.Allowed("")
	}
```

prometherion replied:
I was thinking more about closures, here, as described [here](https://www.calhoun.io/pitfalls-of-context-values-and-how-to-avoid-or-mitigate-them/) although regarding the `context` API

MaxFedotov commented:
@prometherion
I moved logic of getting Tenant for Namespace from `owner_reference` webhook to `utils` package to simplify it's refactoring with #76

prometherion commented:
@MaxFedotov this rebase will be painful, I know :(
If you don't mind, we could try to work on this together: eager to get it merged.

prometherion commented on config/crd/bases/capsule.clastix.io_tenants.yaml:
```diff
@@ -601,7 +605,20 @@ spec:
+                  enum:
+                  - User
+                  - Group
+                  - ServiceAccount
```
Since `ServiceAccount` are namespace scoped resources, I guess this can be dropped.
